### PR TITLE
Trainable mask

### PR DESCRIPTION
**Pull request by YohannPerron**
Add support for trainable mask.

Please also refer to the doc on [contributing](https://lensless.readthedocs.io/en/latest/contributing.html) for more details. 

- [x] Have you run the tests by running `pytest` from the repository root? See the [documentation](https://lensless.readthedocs.io/en/latest/contributing.html#unit-tests) for more details.
- [x] Do you follow the code formatting for this project? This can be done by setting up pre-commit hooks as described in the [documentation](https://lensless.readthedocs.io/en/latest/contributing.html#coding-style).
- [ ] Is there a unit test for the proposed code modification? If the PR addresses an issue, the test should make sure the issue is fixed.
- [x] Are there docstrings? Do they follow the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) style?
- [x] Have you checked that the doc builds properly and that any new RST file has been added to the repo? How to do that is covered in the [documentation](https://lensless.readthedocs.io/en/latest/contributing.html#building-documentation).
- [x] Did you document the proposed change(s) in the CHANGELOG file? It should go under "Unreleased".

Happy PR :smiley:

### CODE REVIEW

**Comment by YohannPerron** [comment]
Not sure where to put the new ``TrainableMask`` in the doc.

**Comment by YohannPerron** [comment]
This is impacted by #82 when simulated a grayscale dataset. This shouldn't be a problem for merging (the issue is probably present even before); but it would be nice to fix #82. I will probably not have the time.

**Comment by ebezzam** [comment]
> Not sure where to put the new `TrainableMask` in the doc.

In the masks page seems appropriate? https://lensless.readthedocs.io/en/latest/mask.html

File: configs/unrolled_recon.yaml [hunk 1]
```diff
@@ -73,18 +80,19 @@ target: "object_plane"    # "original" or "object_plane" or "label"
...
   random_vflip: 0.5
   random_hflip: 0.5
   random_rotate: False
   # these distance parameters are typically fixed for a given PSF
-  # for tape_rgb psf      # for DiffuserCam psf
-  scene2mask: 40e-2       # scene2mask: 10e-2
-  mask2sensor: 4e-3       # mask2sensor: 9e-3
+  # for DiffuserCam psf # for tape_rgb psf     
+  scene2mask: 10e-2     # scene2mask: 40e-2       
+  mask2sensor: 9e-3     # mask2sensor: 4e-3
```
Comment (ebezzam) [inline configs/unrolled_recon.yaml:92]: these are the distances for the DiffuserCam PSF?

File: configs/unrolled_recon.yaml
Comment (ebezzam) [inline configs/unrolled_recon.yaml:1]: instead of editing this file, do you want to create a new one for training masks? and one that's the setting you found best for training the pre- and post-processors with unrolled?

File: lensless/recon/recon.py [hunk 2]
```diff
@@ -404,6 +404,28 @@ def get_image_estimate(self):
         """Get current image estimate as [Batch, Depth, Height, Width, Channels]."""
         return self._form_image()
 
+    def _set_psf(self, psf):
```
Comment (ebezzam) [inline lensless/recon/recon.py:407]: can remove the "_" in beginning of file function? Unless you don't want users to use this? (convention is to begin with "_" if it's a private method)

File: lensless/recon/rfft_convolve.py [hunk 3]
```diff
@@ -57,6 +57,9 @@ def __init__(self, psf, dtype=None, pad=True, norm="ortho", **kwargs):
         self._is_rgb = psf.shape[3] == 3
         assert self._is_rgb or psf.shape[3] == 1
 
+        # save normalization
+        self.norm = norm
```
Comment (ebezzam) [inline lensless/recon/rfft_convolve.py:61]: is this used somewhere?

File: lensless/recon/trainable_mask.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+# #############################################################################
+# trainable_recon.py
```
Comment (ebezzam) [inline lensless/recon/trainable_mask.py:2]: trainable_mask.py

File: lensless/recon/trainable_mask.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+class TrainableMask(metaclass=abc.ABCMeta):
+    """
+    Virtual class for defining trainable masks.
+    The following abstract methods need to be defined:
+    get_psf: getting the PSF of the mask from the mask parameter.
+    project: projecting the mask parameters to a valid space (should be a subspace of [0,1]).
+    """
+
+    def __init__(self, initial_mask, optimizer="Adam", lr=1e-3, **kwargs):
```
Comment (ebezzam) [inline lensless/recon/trainable_mask.py:20]: docstrings, should be straightforward with Copilot ;)

File: lensless/recon/trainable_mask.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        """Abstract method for projecting the mask parameters to a valid space (should be a subspace of [0,1])."""
+        raise NotImplementedError
+
+
+class TrainablePSF(TrainableMask):
+    """
+    Class for defining trainable amplitude masks.
+    """
+
+    def __init__(self, initial_mask, is_rgb=True, optimizer="Adam", lr=1e-3, **kwargs):
```
Comment (ebezzam) [inline lensless/recon/trainable_mask.py:54]: only `is_rgb` has to be defined? as other arguments are in parent class

File: lensless/recon/trainable_mask.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @abc.abstractmethod
+    def project(self):
+        """Abstract method for projecting the mask parameters to a valid space (should be a subspace of [0,1])."""
+        raise NotImplementedError
+
+
+class TrainablePSF(TrainableMask):
+    """
+    Class for defining trainable amplitude masks.
```
Comment (ebezzam) [inline lensless/recon/trainable_mask.py:51]: "... for defining an object that directly optimizes the PSF, without any constraints on what can be realized physically."

File: lensless/utils/dataset.py [hunk 8]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        test_sim = FarFieldSimulator(psf=temp_psf, **simulator.params)
+        assert (
+            test_sim.conv_dim == simulator.conv_dim
+        ).all(), "PSF shape should match simulator shape"
+
+        super(SimulatedDatasetTrainableMask, self).__init__(
+            dataset, simulator, pre_transform, dataset_is_CHW, flip, **kwargs
+        )
+
+    def _get_images_pair(self, index):
```
Comment (ebezzam) [inline lensless/utils/dataset.py:497]: without "_"?

File: lensless/utils/dataset.py [hunk 9]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        # update psf
+        psf = self._mask.get_psf()
+        self.sim = FarFieldSimulator(psf=psf, **self.sim.params)
+        # hacky way to remove the cast to int8 which break differentiability
```
Comment (ebezzam) [inline lensless/utils/dataset.py:501]: remove? what you can do in constructor is to make sure `self.simulator.quantize=False`. (see [original simulator](https://github.com/ebezzam/waveprop/blob/8e9a7e1b199c211f598e73a048083d21e0e2756a/waveprop/simulation.py#L84)). This "turns off" the quantization

File: lensless/utils/simulation.py [hunk 10]
```diff
@@ -73,6 +73,21 @@ def __init__(
...
+            "object_height": object_height,
+            "scene2mask": scene2mask,
+            "mask2sensor": mask2sensor,
+            "sensor": sensor,
+            "output_dim": output_dim,
+            "snr_db": snr_db,
+            "max_val": max_val,
+            "device_conv": device_conv,
+            "random_shift": random_shift,
+            "is_torch": is_torch,
```
Comment (ebezzam) [inline lensless/utils/simulation.py:87]: add "quantize"?

File: lensless/utils/dataset.py [hunk 11]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        dataset_is_CHW=False,
+        flip=False,
+        **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+
+        psf : torch.Tensor
+            PSF to use for simulate. Should be a 4D tensor with shape [1, H, W, C]. Simulation of multi-depth data is not supported yet.
```
Comment (ebezzam) [inline lensless/utils/dataset.py:474]: mask instead of PSF?

File: lensless/utils/dataset.py [hunk 12]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        dataset : torch.utils.data.Dataset
```
Comment (ebezzam) [inline lensless/utils/dataset.py:475]: with :py:class:

File: lensless/utils/dataset.py [hunk 13]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        dataset_is_CHW : bool, optional
+            If True, the input dataset is expected to output images with shape [C, H, W], by default ``False``.
+        flip : bool, optional
+                If True,images are flipped beffore the simulation, by default ``False``..
+        """
+
+        self._mask = mask
+
+        temp_psf = self._mask.get_psf()
+        test_sim = FarFieldSimulator(psf=temp_psf, **simulator.params)
```
Comment (ebezzam) [inline lensless/utils/dataset.py:488]: why not define `self.sim` here? more standard to do inside constructor rather than method

File: lensless/utils/dataset.py [hunk 14]
```diff
@@ -446,3 +446,59 @@ def __init__(
...
+        ).all(), "PSF shape should match simulator shape"
+
+        super(SimulatedDatasetTrainableMask, self).__init__(
+            dataset, simulator, pre_transform, dataset_is_CHW, flip, **kwargs
+        )
+
+    def _get_images_pair(self, index):
+        # update psf
+        psf = self._mask.get_psf()
+        self.sim = FarFieldSimulator(psf=psf, **self.sim.params)
```
Comment (ebezzam) [inline lensless/utils/dataset.py:500]: and why not define a `set_psf` method for `self.sim` rather than recreating an object each time?

File: configs/unrolled_recon.yaml (hunk 1 above)
Comment (YohannPerron) [reply configs/unrolled_recon.yaml:92]: Yes, this is the value you gave me a while ago (I think...)

File: configs/unrolled_recon.yaml
Comment (YohannPerron) [reply configs/unrolled_recon.yaml:1]: Sure

File: lensless/recon/recon.py (hunk 2 above)
Comment (YohannPerron) [reply lensless/recon/recon.py:407]: It is intended to be a private function. For Unrolled and pre/post changing the psf while break the reconstruction.

File: lensless/recon/rfft_convolve.py (hunk 3 above)
Comment (YohannPerron) [reply lensless/recon/rfft_convolve.py:61]: Yes, when changing the psf

File: lensless/recon/trainable_mask.py (hunk 6 above)
Comment (YohannPerron) [reply lensless/recon/trainable_mask.py:54]: Yes, but this makes it clearer in the doc if all the argument are explicit

File: lensless/utils/dataset.py (hunk 13 above)
Comment (YohannPerron) [reply lensless/utils/dataset.py:488]: Because the psf change and we need a new simulator each time

Comment (YohannPerron) [reply lensless/utils/dataset.py:488]: Plus it is already defined by SimulatedFarFieldDataset constructor

File: lensless/utils/dataset.py (hunk 8 above)
Comment (YohannPerron) [reply lensless/utils/dataset.py:497]: It is normal! see DualDataset

File: lensless/utils/dataset.py (hunk 14 above)
Comment (YohannPerron) [reply lensless/utils/dataset.py:500]: It would be a change for Waveprop. If you are willing to do it, sure! It would be better. I am not familiar enough with waveprop to do it myself without breaking a thing or two ^^

File: lensless/utils/dataset.py (hunk 9 above)
Comment (YohannPerron) [reply lensless/utils/dataset.py:501]: Fixed

File: lensless/utils/simulation.py (hunk 10 above)
Comment (YohannPerron) [reply lensless/utils/simulation.py:87]: Done

File: configs/unrolled_recon.yaml (hunk 1 above)
Comment (ebezzam) [reply configs/unrolled_recon.yaml:92]: ok just to be sure
